File: src/uva/Leetcode00023.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Leetcode00023 {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> p = new PriorityQueue<ListNode>(10, new Comparator<ListNode>()
            {
                public int compare(ListNode o1, ListNode o2) {
                    return o1.val - o2.val;
                }
            }
        );
        for (ListNode l : lists) {
            if (l != null) {
                p.add(l);
            }
        }
        ListNode prehead = new ListNode(0);
        ListNode node = prehead;
        while (!p.isEmpty()) {
            ListNode next = p.poll();
            if (next.next != null) {
                p.add(next.next);
            }
            node.next = next;
            node = node.next;
        }
        return prehead.next;
    }
}

File: src/main/Leetcode00023.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
 
 class Heap <T:IntegerType> {
    var items = [T](count:1, repeatedValue:0)
    var n = 0
    func push(t : T) {
        items.append(t)
        n += 1
        swim(n)
    }
    func pop() -> T? {
        if n == 0 {
            return nil
        } else if n > 1 {
            swap(&items[1], &items[n])
        }
        n -= 1
        sink(1)
        return items.removeLast()
    }
    func count() -> Int {
        return n
    }
    func sink(index : Int) {
        var i = index
        while i*2 <= n {
            var j = i*2
            if j + 1 <= n && items[j+1] < items[j] {
                j += 1
            }
            if items[i] < items[j] {
                break
            }
            swap(&items[i], &items[j])
            i = j
        }
        
    }
    func swim(index : Int) {
        var i = index
        while(i > 1 && items[i] < items[i/2]) {
            swap(&items[i], &items[i/2])
            i = i/2
        }
    }
}

class Solution {
    func mergeKLists(lists: [ListNode?]) -> ListNode? {
        var heap = Heap<Int>()
        for list in lists {
            var node = list
            while node != nil {
                print(node!.val)
                heap.push(node!.val)
                node = node!.next
            }
        }
        var prehead = ListNode(0), node : ListNode? = prehead
        while let val = heap.pop() {
            print(val)
            node!.next = ListNode(val)
            node = node!.next
        } 
        return prehead.next
    }
}
